### PR TITLE
[Core] Improve and clean condition number utility

### PR DESCRIPTION
**Pull request by loumalouomega**
This clean ups the condition number utility. Clean ups and improves the eigen_solver_factory. Corrects ands clean ups teh use of the condition n umber utility in the ContactStructuralMechanics app (I haven't use it since before pybind11 and it was broken)

### CODE REVIEW

**Comment by loumalouomega** [comment]
@inigolcanalejo I know you used the stuff, in theory this odes not change anything relevant

**Comment by inigolcanalejo** [comment]
Thanks @loumalouomega ! Yes, I have been (and I am still) using your utility to check the condition number of the system matrix in the potential flow app. I have been using it even after pybind11 and was still working fine for me.

I have tried with this branch and it does run much faster! :+1: The changes are ok with me. Since there are changes in the core I leave the approval for @philbucher.

**Comment by loumalouomega** [comment]
> 
> 
> Thanks @loumalouomega ! Yes, I have been (and I am still) using your utility to check the condition number of the system matrix in the potential flow app. I have been using it even after pybind11 and was still working fine for me.
> 
> I have tried with this branch and it does run much faster! 👍 The changes are ok with me. Since there are changes in the core I leave the approval for @philbucher.

The difference in the speed is because now it takes the fastest linear solver available by default instead of defining by default the SkylineLU (the slowest of the linear solvers). The thing of the pybind is because in my app I use the pointer to the utility, and you need to tell explicitly to pybind11 to export it as a pointer

File: kratos/python_scripts/eigen_solver_factory.py [hunk 1]
```diff
@@ -1,31 +1,69 @@
...
+        if kratos_utils.CheckIfApplicationsAvailable("ExternalSolversApplication"):
+            eigen_solver = ExSA.FEASTSolver(settings, linear_solver)
+        else:
+            raise Exception("ExternalSolversApplication not available")
     else:
         raise Exception("Solver type not found. Asking for :" + solver_type)
 
     return eigen_solver
+
+def CreateLinearSolver(settings):
```
Comment (philbucher) [inline kratos/python_scripts/eigen_solver_factory.py]:
```suggestion
def CreateDirectLinearSolver(settings):
```

**Review by philbucher — CHANGES_REQUESTED** [review]
Thanks a lot, blocking bcs I want to take a closer look but that will be end of the week
Sorry for the delay, will do it asap

File: kratos/python_scripts/eigen_solver_factory.py (hunk 1 above)
Comment (loumalouomega) [reply kratos/python_scripts/eigen_solver_factory.py]: We should move this to a common utility perhaps, it is basically copied from the mechanical_solver.py

Comment (philbucher) [reply kratos/python_scripts/eigen_solver_factory.py]: I also thought this
I suggest to move it to the python-lin-solver-factory

Comment (loumalouomega) [reply kratos/python_scripts/eigen_solver_factory.py]: I was thinking the same

File: applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_solver.py [hunk 2]
```diff
@@ -16,6 +16,12 @@
 from KratosMultiphysics import auxiliary_solver_utilities
 import KratosMultiphysics.kratos_utilities as kratos_utils
 
+if kratos_utils.CheckIfApplicationsAvailable("EigenSolversApplication"):
+    import KratosMultiphysics.EigenSolversApplication as EiSA
+
+if kratos_utils.CheckIfApplicationsAvailable("ExternalSolversApplication"):
+    import KratosMultiphysics.ExternalSolversApplication as ExSA
```
Comment (philbucher) [inline applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_solver.py]: importing bcs you might need is is not allowed. The factory imports it if necessary
(by the way we had problems with this)

File: kratos/python/add_utilities_to_python.cpp [hunk 3]
```diff
@@ -331,7 +331,7 @@ void AddUtilitiesToPython(pybind11::module &m)
     InputGetConditionNumber ThisGetConditionNumber = &ConditionNumberUtility::GetConditionNumber;
     DirectGetConditionNumber ThisDirectGetConditionNumber = &ConditionNumberUtility::GetConditionNumber;
 
-    py::class_<ConditionNumberUtility>(m,"ConditionNumberUtility")
+    py::class_<ConditionNumberUtility,ConditionNumberUtility::Pointer>(m,"ConditionNumberUtility")
```
Comment (philbucher) [inline kratos/python/add_utilities_to_python.cpp:334]: why is this needed?
I usually use it without the `Pointer` for simple cases such as this

File: kratos/python_scripts/eigen_solver_factory.py [hunk 4]
```diff
@@ -1,30 +1,43 @@
...
+import KratosMultiphysics as KM
+
+from KratosMultiphysics import python_linear_solver_factory as linear_solver_factory
+import KratosMultiphysics.kratos_utilities as kratos_utils
+
+if kratos_utils.CheckIfApplicationsAvailable("EigenSolversApplication"):
+    import KratosMultiphysics.EigenSolversApplication as EiSA
+
+if kratos_utils.CheckIfApplicationsAvailable("ExternalSolversApplication"):
+    import KratosMultiphysics.ExternalSolversApplication as ExSA
```
Comment (philbucher) [inline kratos/python_scripts/eigen_solver_factory.py]: same as above

File: kratos/python_scripts/python_linear_solver_factory.py [hunk 5]
```diff
@@ -168,3 +168,28 @@ def ConstructSolver(configuration):
...
             "Constructing a regular (non-complex) linear-solver")
         return KM.LinearSolverFactory().Create(configuration)
+
+def CreateDirectLinearSolver(settings):
+    if not settings.Has("linear_solver_settings"):
+        settings.AddEmptyValue("linear_solver_settings")
+    linear_solver_configuration = settings["linear_solver_settings"]
+    if linear_solver_configuration.Has("solver_type"): # user specified a linear solver
+        return ConstructSolver(linear_solver_configuration)
+    else:
```
Comment (philbucher) [inline kratos/python_scripts/python_linear_solver_factory.py]: this method should only do the things after `else`
Otherwise it could also create an iterative solver

**Review by philbucher — COMMENTED** [review]
there are some issues, I am fixing it myself now

File: kratos/python/add_utilities_to_python.cpp (hunk 3 above)
Comment (loumalouomega) [reply kratos/python/add_utilities_to_python.cpp:334]: In my application I pass it as an argument, so I need to be a pointer, in general is not necessary

File: kratos/python_scripts/eigen_solver_factory.py (hunk 4 above)
Comment (loumalouomega) [reply kratos/python_scripts/eigen_solver_factory.py]: I trust you, I am a little lost with the imports

File: kratos/python/add_utilities_to_python.cpp (hunk 3 above)
Comment (philbucher) [reply kratos/python/add_utilities_to_python.cpp:334]: ok, understand

File: kratos/python_scripts/python_linear_solver_factory.py (hunk 5 above)
Comment (loumalouomega) [reply kratos/python_scripts/python_linear_solver_factory.py]: True (I think that the only limitation is the linear solver, the iterative would be fine I think)

File: kratos/python_scripts/eigen_solver_factory.py (hunk 4 above)
Comment (philbucher) [reply kratos/python_scripts/eigen_solver_factory.py]: Basic guide:
- Import the things you need for sure at the top of the file
- Do not import things you might not need (e.g. the solver-apps) => those will be imported by the corresponding factories

**Review by philbucher — DISMISSED** [review]
now should be ok, can you have a look @loumalouomega ?

File: kratos/python_scripts/python_linear_solver_factory.py [hunk 6]
```diff
@@ -168,3 +168,27 @@ def ConstructSolver(configuration):
...
+    # Using a default linear solver (selecting the fastest one available)
+    if kratos_utils.CheckIfApplicationsAvailable("EigenSolversApplication"):
+        from KratosMultiphysics import EigenSolversApplication
+    elif kratos_utils.CheckIfApplicationsAvailable("ExternalSolversApplication"):
+        from KratosMultiphysics import ExternalSolversApplication
+
+    linear_solvers_by_speed = [
+        "pardiso_lu", # EigenSolversApplication (if compiled with Intel-support)
+        "sparse_lu",  # EigenSolversApplication
+        "pastix",     # ExternalSolversApplication (if Pastix is included in compilation)
```
Comment (loumalouomega) [inline kratos/python_scripts/python_linear_solver_factory.py:180]: IMO pastix is or the first or the second faster solver (we must update the solver BTW :P )

**Comment by loumalouomega** [comment]
OK

File: kratos/python_scripts/python_linear_solver_factory.py (hunk 6 above)
Comment (philbucher) [reply kratos/python_scripts/python_linear_solver_factory.py:180]: I doubt it is faster than pardiso
at least here everyone switched from pastix to pardiso since it became available

**Review by philbucher — APPROVED** [review]

File: kratos/python_scripts/python_linear_solver_factory.py (hunk 6 above)
Comment (loumalouomega) [reply kratos/python_scripts/python_linear_solver_factory.py:180]: Perhaps only in my ill-conditioned contact problem

Comment (philbucher) [reply kratos/python_scripts/python_linear_solver_factory.py:180]: :D

Why again was it necessary to update it?
I remember there was an issue but I forgot the details

Comment (loumalouomega) [reply kratos/python_scripts/python_linear_solver_factory.py:180]: Because right now we are using the legacy version. The last version is faster and better, and MPI compatible BTW, but refactors everything, so the adaptation must be done from scratch

Comment (philbucher) [reply kratos/python_scripts/python_linear_solver_factory.py:180]: oh, that sound like quite some effort :/

I would be curious to try the MPI-version

Comment (loumalouomega) [reply kratos/python_scripts/python_linear_solver_factory.py:180]: http://pastix.gforge.inria.fr/files/README-txt.html according to the last papers published we speak of 4x-6x faster

Comment (philbucher) [reply kratos/python_scripts/python_linear_solver_factory.py:180]: impressive

we cannot include it permantently bcs of the license right?

Comment (loumalouomega) [reply kratos/python_scripts/python_linear_solver_factory.py:180]: yep, cecil is a strange licence, is basically a LGPL by INRIA
